Implementation of LOA single-tree

ZedongPeng opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -370,20 +376,43 @@ def _add_loa_configs(CONFIG):
+    ))
+    CONFIG.declare('reduce_level_coef', ConfigValue(
+        default=False,
+        description='Whether to reduce level coefficient in LOA single tree when projection problem is infeasible',
```
I do not understand why this would solve the infeasibility in the projection problem. By reducing the coefficient by 1/2 you basically make the solutions close to the incumbent feasible.

ZedongPeng replied:
We can talk about this in detail at the meeting.

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -370,20 +376,43 @@ def _add_loa_configs(CONFIG):
+    ))
+    CONFIG.declare('use_master_incumbent', ConfigValue(
+        default=False,
+        description='Whether to use the incumbent solution of master problem in LOA single tree when projection problem is infeasible',
```
In the case it is not the master incumbent, we use the incumbent of the whole tree?

ZedongPeng replied:
Yes

bernalde commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -66,10 +66,10 @@ def add_oa_cuts(target_model, dual_values, solve_data, config,
                     and (linearize_active and abs(constr.uslack()) < config.zero_tolerance) \
                         or (linearize_violated and constr.uslack() < 0) \
-                        or (config.linearize_inactive and constr.uslack() > 0):
+                        or (config.linearize_inactive and constr.uslack() > 0)) or (constr.name == 'MindtPy_utils.objective_constr' and constr.has_ub()):
```
This is to always add the linearization for the epigraph of the objective, right? Maybe we can add a comment mentioning it.

ZedongPeng replied:
Yep. OK.

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -207,6 +215,14 @@ def algorithm_should_terminate(solve_data, config, check_cycling):
+        if solve_data.UB - solve_data.LB <= config.relative_bound_tolerance * (abs(solve_data.UB if solve_data.objective_sense == minimize else solve_data.LB) + 1E-10):
+            config.logger.info(
+                'MindtPy exiting on bound convergence. '
+                '(UB: {} - LB: {})/ (1e-10+|bestinteger|:{}) <= relative tolerance: {}'.format(solve_data.UB, solve_data.LB, abs(solve_data.UB if solve_data.objective_sense == minimize else solve_data.LB), config.relative_bound_tolerance))
```
Very elegant! We should add the formula in the description of the parameter.

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -98,26 +96,28 @@ def solve_master(solve_data, config, fp=False, regularization_problem=False):
         masteropt._solver_model.set_error_stream(None)
     mip_args = dict(config.mip_solver_args)
-    if config.mip_solver in {'cplex', 'gurobi'}:
+    if config.mip_solver in {'cplex','cplex_persistent', 'gurobi','gurobi_persistent'}:
```
Does this work with Gurobi persistent already?

ZedongPeng replied:
For single-tree, not yet. Gurobi persistent can be used as the mip solver in the multi-tree version like any other mip solvers.

bernalde replied:
Good, that was not clear to me. How hard would it be to extend the single tree implementation to Gurobi?

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -380,6 +384,9 @@ def setup_master(solve_data, config, fp, regularization_problem):
     MindtPy.del_component('mip_obj')
+    if regularization_problem and config.single_tree:
+        MindtPy.del_component('loa_proj_mip_obj')
+        MindtPy.cuts.del_component('obj_reg_estimate')
```
Why are these components deleted here?

ZedongPeng replied:
These components are updated in each iteration(each incumbent solution in single-tree). If we define these components twice, pyomo will report a warning. Therefore, I delete these components before updating it.

bernalde commented on pyomo/contrib/mindtpy/single_tree.py:
```diff
@@ -493,6 +500,48 @@ def __call__(self):
 
+        # regularization is activated after the first feasible solution is found.
+        if config.add_regularization is not None and solve_data.best_solution_found is not None:
+            # the master problem might be unbounded, regularization is activated only when a valid bound is provided.
```
Perfect!

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -421,40 +419,72 @@ def set_solver_options(opt, solve_data, config, type, regularization=False):
         if regularization == True:
             if solver_name == 'cplex':
-                opt.options['mip limits populate'] == 10
+                opt.options['mip limits populate'] = 10
```
This limit should be a MindtPy option, let's avoid setting arbitrary parameters as constants. Similarly with the mipgap/optcr

ZedongPeng replied:
What `optcr` value should we set for NLP solvers?

bernalde replied:
I do not care too much about the optcr for NLP solvers, while I do care for the MIP. This since for example in GAMS that value is by default 0.1. We do have optimality tolerances in MindtPy, so we can apply those same tolerances for the subproblems.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -421,40 +419,72 @@ def set_solver_options(opt, solve_data, config, type, regularization=False):
-        elif type == 'nlp':
+        elif solver_type == 'nlp':
             opt.options['add_options'] = ['option reslim=%s;' % remaining]
+            if config.nlp_solver_args['solver'] in {'ipopt', 'ipopth', 'msnlp', 'baron'}:
```
All the things down here are absolutely wonderful!

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -325,7 +323,7 @@ def generate_lag_objective_function(model, setpoint_model, config, solve_data, d
+        first_order_term = sum(float(jac_lag[nlp.get_primal_indices([temp_var])[0]]) * (var - temp_var.value) for var,
                                temp_var in zip(model.MindtPy_utils.variable_list[:-1], temp_model.MindtPy_utils.variable_list[:-1]) if temp_var.name in nlp_var)
 
         # Implementation 2
```
Are you still considering the 2 implementations for the Hessian of the Lagrangian? If only one is used leave it commented.

bernalde commented on pyomo/contrib/mindtpy/single_tree.py:
```diff
@@ -493,6 +500,48 @@ def __call__(self):
+                    if config.reduce_level_coef:
+                        config.level_coef = config.level_coef / 2
+                        master_mip, master_mip_results = solve_master(solve_data, config, regularization_problem=True)
+                        config.level_coef = config.level_coef * 2
```
I'm confused here. Do you restore the coefficient after solving the regularization problem in case that it is infeasible? This is done for a single iteration, but I would expect that if the regularization problem fails once with a coefficient, it will fail in the subsequent iterations. In fact, it shouldn't fail in the first place since we are initializing the regularization problem with the master solution. At the very least let's use the "reduction coefficient" as a parameter (here it is 2) so that anyone can play with it. This is one point we should mainly discuss (together with the fake bounds)

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -370,20 +376,43 @@ def _add_loa_configs(CONFIG):
             config.projection_mip_threads = config.threads
             config.logger.info('Set projection_mip_threads equal to threads')
+        if config.single_tree:
+            if not (config.use_fake_bound or config.reduce_level_coef or config.use_master_incumbent):
```
It took me a minute to understand the logic here. Add a comment explaining the reasoning behind it

bernalde reviewed (commented):
Very nice work @ZedongPeng . I think that the code looks very professional and most of my comments are just requesting either a little documentation or adding extra parameters to MindtPy.
There are two main points that we need to talk about.
1. The fake bounds idea for single tree appears to be reasonable to me. It is a little surprising that it does not work without this option since we are computing the optimistic (max) bound for the LB in minimization using the incumbent information, but if it works, it works!
2. The coefficient reduction trick to make the regularization problem feasible in counterintuitive. Reducing the coefficient does increase the feasible region's size, but in the direction towards the incumbent. Since we are initializing with the master solution, this should be feasible by default, and moving the coefficient should not make a difference (unless it is a thing of numerical tolerances).
In any case, things look good. Let's complete the numerical experiments and see if things keep failing (which I do not expect).

jsiirola committed:
Move to using 'yield from' in kernel generators

jsiirola committed:
Adding a pypy build to the branch tests

jsiirola committed:
Additional moves to using 'yield from' in core, repn

jsiirola committed:
Merge branch 'main' into kernel-generators

jsiirola committed:
Clean up logic for skipping extensions on slim builds

jsiirola committed:
Enable AmplInterface.available() even in the absence of numpy

jsiirola committed:
Skip some incidence_analysis tests when not AmplInterface.available()

jsiirola committed:
Switch piecewise solve tests to use gurobi

jsiirola committed:
Update baseline string names (quoting strings)

jsiirola committed:
Fix regression importing numpy

emma58 committed:
Using fbbt to calculate M for linear expressions too

emma58 committed:
Remove extra space in error message

michaelbynum committed:
fbbt support for linear expression

ZedongPeng committed:
add unit_test.py

ZedongPeng committed:
increase code coverage

michaelbynum committed:
Merge pull request #1994 from michaelbynum/fbbt_linear_expr

fbbt support for linear expression

jsiirola committed:
Catch non-standard file.write() implementations

Catch non-standard file object implementations that do not return the
number of characters written as the result of `write()`.  Fixes #1997.

jsiirola committed:
Update expected test output

jsiirola committed:
Merge pull request #1991 from jsiirola/testing-fixes

Testing fixes and updates

jsiirola committed:
Add explanatory comment

jsiirola committed:
Merge pull request #1999 from jsiirola/tee-ipython-fix

Catch non-standard `file.write()` implementations

emma58 committed:
Merge branch 'main' into switch-bigm-to-fbbt

michaelbynum committed:
Merge pull request #2000 from emma58/switch-bigm-to-fbbt

Switch bigm's M value calculation to always use fbbt

jsiirola committed:
Implement redirect_fd for robust redirection of file descriptors

jsiirola committed:
Move cyipopt to using common.tee.redirect_fd

jsiirola committed:
Fixing typo (undefined variable)

jsiirola committed:
Fixing doc typo

jsiirola committed:
Add redirect_fd documentation; remove enable flag

jsiirola committed:
Fix redirection of cyipopt stdout to sys.stdout

jsiirola committed:
Fixing typo (copy/paste error)

jsiirola committed:
Removing deprecated import of 'ipopt' module

jsiirola committed:
Update cyipopt interface to use the 1.0.0 API

jsiirola committed:
NFC: fix doc typo

jsiirola committed:
Ensure test output is flushed (fixes failures on pypy)

blnicho committed:
A few minor changes in GDPopt and examples to avoid implicit conversion warnings

blnicho committed:
Updating the CHANGELOG in preparation for the 6.0.1 release

jsiirola committed:
Silence deprecation warnings in examples/gdp/stickies.py

blnicho committed:
Merge pull request #2001 from jsiirola/capture-fd

Improve management of stdout in cyipopt solver

blnicho committed:
Merge branch 'main' into finalize-release-6.0.1

blnicho committed:
Merge pull request #2002 from blnicho/fix-gdpopt-examples

Fixes to GDPopt and examples to avoid implicit conversion warnings

blnicho committed:
More updates to the CHANGELOG for the 6.0.1 release

blnicho committed:
Merge branch 'main' into finalize-release-6.0.1

jsiirola committed:
Updating version to 6.0.1

blnicho committed:
Merge pull request #2003 from blnicho/finalize-release-6.0.1

Finalize release 6.0.1

jsiirola committed:
Resetting main for development (6.0.2.dev0)

blnicho committed:
Merge pull request #2004 from blnicho/finalize-release-6.0.1

Resetting main for development (6.0.2.dev0)

mrmundt committed:
Changing implementation of keys, values, items in IndexedComponent

mrmundt committed:
Attempting use of iter option

mrmundt committed:
Converting test generator to list

mrmundt committed:
Changing cast

mrmundt committed:
Converting linear_coefs to a list

mrmundt committed:
Merge pull request #2007 from mrmundt/iterators

Yield Iterator in `IndexedComponent`

jsiirola committed:
Allow objects other than strings as attempt_import(error_message)

jsiirola committed:
LoggingIntercept: log to an implicit StringIO if no buffer provided

jsiirola committed:
Remove explicit dependency on parameterized in APPSI

jsiirola committed:
cyipopt.available() should always return a bool

jsiirola committed:
Remove hard dependency on nose

jsiirola committed:
Ensure direct/persistent solver .available() always returns bool

jsiirola committed:
error_message=None should be handled the same as an empty string

jsiirola committed:
Capture stdout when creating gurobi model to test license

jsiirola committed:
Add a backport of the Python 3.7 nullcontext

jsiirola committed:
Suppress executable not found error in SystemCallSolver.available()

jsiirola committed:
Update Gurobi direct/persistent to leverage attempt_import

Also, suppress stdout when testing available()

jsiirola committed:
Update xpress_direct to leverage attempt_import

Also suppress stdout within available()

jsiirola committed:
Add a standalone Pyomo test driver

This script exercises a writer, creates (almost) all solvers and checks
their availability (ensuring that `available(False)` is silent: no
stdout, no logged messages), creates all (nondeprecated) transformations
and checks that they can be created silently.

jsiirola committed:
Adding '[DEPRECATED]' to deprecated transformation documentation

jsiirola committed:
Add a test that Pyomo installs/runs in a "bare" Python environment

jsiirola committed:
Fixing typo (trailing ")

jsiirola committed:
Fix parameterized import

jsiirola committed:
Simplify import

jsiirola committed:
Remove instance-level definition of _version_major

jsiirola committed:
Fix instance definition of _version attribute

jsiirola committed:
Fix masking _version in gurobi, xpress direct solvers

jsiirola committed:
Fix typo (calling method on class)

jsiirola committed:
Remove pyutilib from branch tests; add bare to pr tests

jsiirola committed:
Support using the TicTocTimer as a context manager

jsiirola committed:
Relax timing tests on pypy

jsiirola committed:
Define 'tests' and 'docs' install dependencies

jsiirola committed:
Switch RTD to using a .readthedocs.yaml configuration

Dependencies are now set in setup.py and installed using
`pip install '.[docs]'`

jsiirola committed:
Relaxing the timing resolution

jsiirola committed:
Adding additional dependencies needed for RTD

jsiirola committed:
Switch timing tests to using the perf_counter (for accuray)

jsiirola committed:
Improve robustness of timing test

michaelbynum committed:
Merge pull request #2012 from jsiirola/timing-updates

Support using TicTocTimer as a context manager

jsiirola committed:
Adding mixxing copyright; fixing comment typos

jsiirola committed:
changing step order in bare-env test

jsiirola committed:
Merge pull request #2011 from jsiirola/dependency-andsolverio-cleanup

Clean up dependencies and `solver.available()` IO

blnicho committed:
Merge pull request #1984 from jsiirola/kernel-generators

[WIP] Use `yield from` in `pyomo.core`

jsiirola committed:
Adding [optional] installation target

jsiirola committed:
Update patch to suppress building benchmark and nl-example targets

These do not cleanly build on newer compilers and are not needed for libasl.

jsiirola committed:
Switch GHA to ubuntu-latest

jsiirola committed:
(debug) add pypy to the GHA branch build

jsiirola committed:
Restore spaces in patch (required for windows patch)

jsiirola committed:
Restore branch linux/3.9 build

jsiirola committed:
Further relax timing targets on pypy

jsiirola committed:
(debug) make parmest more verbose

jsiirola committed:
(debug) print out PATH before running tests

jsiirola committed:
(debug) print out .profile

jsiirola committed:
Move local runner paths before conda

jsiirola committed:
Update PATH in both .profile & .bash_profile

jsiirola committed:
Attempt to correct escaping

jsiirola committed:
ensuring a CR after conda config

jsiirola committed:
Removing debugging

jsiirola committed:
Synchronizing branch/pr workflows

jsiirola committed:
removing debugging

jsiirola committed:
Add 'setup.py dependencies' command

jsiirola committed:
Switch branch builder to use setup.py for dependencies

jsiirola committed:
Merge pull request #2015 from jsiirola/pynumero-build-updates

Update GHA to ubuntu-latest; resolve ampl/mp build error

jsiirola committed:
Fixing typo

jsiirola committed:
Fix catch of pip install

jsiirola committed:
Add fallback for older setuptools

jsiirola committed:
Invoke pip with PYTHON_EXE

jsiirola committed:
Correct fallback distutils import

jsiirola committed:
Do not use 'conda develop': it does not appear to handle console_scripts

jsiirola committed:
Allow empty 'extrs' argument

jsiirola committed:
Filter out packages not available through conda

jsiirola committed:
Always use 'pip install .[extras]' to install pyomo

jsiirola committed:
(debug) quoting dependencies

jsiirola committed:
(debug) pip command arguments

jsiirola committed:
remove use of pip -e

jsiirola committed:
Move pyomo dependency installs earlier

jsiirola committed:
call python directly

jsiirola committed:
Update dependency management

jsiirola committed:
Defer dependencies until after Python is set up

jsiirola committed:
typo: quoting

jsiirola committed:
remove package quoting

jsiirola committed:
Merge branch 'main' into setup-updates

jsiirola committed:
Explicitly list packages that are only available on PyPI

jsiirola committed:
resynchronixe pr/main builder

jsiirola committed:
typo: 'pypi'

jsiirola committed:
Remove extra backquote

jsiirola committed:
Capitalize DependenciesCommand class; clarify documentation

jsiirola committed:
Fix typos, remove unneeded commented out code

jsiirola committed:
Resolve minor doctest errors; add +IGNORE_RESULT option

jsiirola committed:
Make more optional dependencies available on pypy

jsiirola committed:
avoid direct execution of 'pip'

jsiirola committed:
Avoiding casadi under pypy

jsiirola committed:
Avoiding pandas under pypy

jsiirola committed:
Resyncing build scripts

jsiirola committed:
Bumping cache version

jsiirola committed:
Avoiding numdifftools, pandas under pypy

jsiirola committed:
Install testing dependencies when processing coverage

jsiirola committed:
Move constrained_layout tests to 'expensive'; add expensive coverage

mrmundt committed:
Merge pull request #2017 from jsiirola/setup-updates

Embed dependencies in setup.py

mrmundt committed:
Merge pull request #2018 from jsiirola/constrained-layout

Move constrained_layout tests to 'expensive' suite

jsiirola committed:
Merge branch 'main' into new-lp-nlp
